src/net_processing.cpp

```diff
@@ -2842,6 +2848,31 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         return;
     }
 
+    // Feature negotiation of disabletx must happen between VERSION and VERACK.
+    if (msg_type == NetMsgType::DISABLETX) {
+        if (pfrom.fSuccessfullyConnected) {
+            // Disconnect peers that send a disabletx after VERACK.
+            LogPrint(BCLog::NET, "disabletx received after verack from peer=%d; disconnecting\n", pfrom.GetId());
+            pfrom.fDisconnect = true;
+            return;
+        }
+        if (pfrom.IsFullOutboundConn()) {
+            // If we picked an outbound for transaction relay and it sends us a
+            // disabletx, we want to find another peer.
+            pfrom.fDisconnect = true;
+            return;
+        }
+        if (pfrom.m_tx_relay != nullptr && WITH_LOCK(pfrom.m_tx_relay->cs_filter, return pfrom.m_tx_relay->fRelayTxes)) {
+            // Can't send a disabletx message if they didn't turn off
+            // transaction relay in the version message.
```

Comment:
I don't think this interaction with BIP37 is described in the new BIP. The following case "If a node sets the transaction relay field in the version message to a peer to true it MUST NOT send a disabletx message" isn't covered ?

Or do you consider our software to be more severe than the BIP and another implementation can decide to override BIP 37 tx-relay in version message ? It's maybe what you meaned with "This proposal is compatible with, but independent of, BIP 37. "

Reply:
Receiver-side, we can enforce a _consistency_ requirement between `fRelay` and DISABLETX to detect buggy upgraded peers signaling differing relay type preferences.

Sender-side, if we're connected to a non-upgraded peer (`nVersion` < 70017), we should send a `fRelay`, otherwise they can't discover our relay preferences. If we're connected to an upgraded peer, we can ignore setting `fRelay` in our version message and just send a DISABLETX.

I don't think we have specific considerations to avoid disruption BIP37-related logic, in the sense of upgrading the connection type to full-relay if they signal `fRelay=false`. Doing so would break the purpose of our block-relay-only strategy. Actually, that's what we already do, at VERSION message reception, we ignore `fRelay` if peer doesn't have a `m_tx_relay` struct.

Network-wise,, it'll be safe to deprecate BIP37-related logic on our side (i.e don't bother sending `fRelay` even to non-upgraded peers) only once BIP338 is widely deployed. Otherwise, legacy peers will be automatically disconnected when they're picked up as block-relay-only, thus impoverishing their connection graph (only able to connect as full-relay inbound slots).

Note, I think the BIP change should waiwe about the omission case, otherwise it make a requirement to implement BIP60 only for the sake of sending a `fRelay=false`, and thus create an interdependency
between both standards ?

Reply:
> Sender-side, if we're connected to a non-upgraded peer (nVersion < 70017), we should send a fRelay, otherwise they can't discover our relay preferences. If we're connected to an upgraded peer, we can ignore setting fRelay in our version message and just send a DISABLETX.

I think we still want to send fRelay=false, because we don't know that an upgraded peer has implemented the BIP -- the p2p version bump is just to signify which p2p messages are considered valid.  (For instance, some other new p2p message may be introduced for some other reason, which might cause someone to bump the p2p version they are advertising -- even if they haven't added support for BIP 338.)

So I think given that, we might as well treat fRelay as a required part of the spec for software that is implementing `disabletx`.

Reply:
Thanks to recall that p2p version advertising doesn't imply support of the well-known associated set of messages with that version number.

If so, I agree that if we're connected to an upgraded peer, we should send `fRelay=false` as an alternative mechanism to disable transaction relay messages on the connection. However, if we want to deprecate `fRelay` in the future once BIP338 is deployed, I think we should not treat `fRelay` as a required part of the spec.

To the best of my understanding, the proposed BIP338 draft is ambiguous as it sounds to make a requirement of a `disabletx` receiving peer to support first BIP37 ("(provided that BIP 37 or BIP 60 has been implemented") though at the same time explicitly says those standards are independent ("This proposal is compatible with, but independent of, BIP 37. ").

I would favor to make the BIP37 deprecation option more explicit in BIP338 (i.e remove "provided that BIP37...") and that dual-support is an implementation choice, not a spec requirement.

That said, this comment is not a substantial blocker to me to advance this work forward.